### PR TITLE
Added firmware simulation boilerplate

### PR DESCRIPTION
**Pull request by MathewMacDougall**
### Description

Added wrapper classes for the simulator Physics objects so we can interface with our firmware and eventually simulate it. Added lots of boilerplate to these classes, but the functions we provide to the firmware still need to be implemented (will be my next PR).

Also gave the robots the correct mass in simulation.

### Testing Done

Unit tested the wrapper classes to verify that our strategy of "switching" concrete objects behind the scenes works as expected to allow us to control multiple objects and manage "multiple" instances of firmware.

### Resolved Issues

This didn't have an explicit ticket yet but is part of both `#768` and `#883` (formatted to github doesn't close them automatically)

### Length Justification

Lots of boilerplate and function stubs so shouldn't be as bad to review, and was easier to do it all at once to make sure it worked.

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/firmware/main/app/world/firmware_ball.h
Comment (akhilveeraghanta) [inline src/firmware/main/app/world/firmware_ball.h]: nit: revert

File: src/software/backend/simulation/physics/physics_simulator.h
Comment (akhilveeraghanta) [inline src/software/backend/simulation/physics/physics_simulator.h:64]: nit: remove TODO

File: src/software/backend/simulation/physics/physics_simulator.h
Comment (akhilveeraghanta) [inline src/software/backend/simulation/physics/physics_simulator.h:68]: ifeelbadevenpointingthisoutnit: can we join physics and world to PhysicsWorld it was hard to read separately

**Review by akhilveeraghanta — COMMENTED** [review]
LGTM coming together nicely, just a few super minor things

File: src/software/backend/simulation/physics/box2d_util.cpp
Comment (jared-paul) [inline src/software/backend/simulation/physics/box2d_util.cpp]: NIT: "when they are created"

File: src/software/backend/simulation/physics/physics_simulator.h
Comment (jared-paul) [reply src/software/backend/simulation/physics/physics_simulator.h]: I'm working on this right now

**Review by jared-paul — CHANGES_REQUESTED** [review]
LGTM! 
Will hopefully have the PhysicsWorld pushed by Wednesday, sorry it's been hectic with course changes.

File: src/software/backend/simulation/physics/physics_simulator.h
Comment (akhilveeraghanta) [reply src/software/backend/simulation/physics/physics_simulator.h:64]: nvm mb: read it as std::weak_ptr<PhsyicsRobot>

File: src/software/backend/simulation/simulator_ball.cpp
Comment (jonathanlew) [inline src/software/backend/simulation/simulator_ball.cpp]: Is this necessary since it's already been declared in the header? Is this being redeclared and hiding the variable in the header by the same name?

**Review by jonathanlew — CHANGES_REQUESTED** [review]
I'm not sure if it's documented somewhere, but I'd appreciate a comment explaining why exactly we're using weak_ptrs.

File: src/software/backend/simulation/physics/physics_simulator.h
Comment (MathewMacDougall) [reply src/software/backend/simulation/physics/physics_simulator.h:68]: Yea @jared-paul is doing PhysicsWorld, will be merged separately. Once it is I'll join stuff up.

File: src/firmware/main/app/world/firmware_ball.h
Comment (MathewMacDougall) [reply src/firmware/main/app/world/firmware_ball.h]: fixed

File: src/software/backend/simulation/physics/box2d_util.cpp
Comment (MathewMacDougall) [reply src/software/backend/simulation/physics/box2d_util.cpp]: fixed

**Comment by MathewMacDougall** [comment]
> I'm not sure if it's documented somewhere, but I'd appreciate a comment explaining why exactly we're using weak_ptrs.

This is explained in the javadoc on line `86` of `physics_simulator.h`, since it's the "source" of why `weak_ptrs` are used

File: src/shared/constants.h
Comment (MathewMacDougall) [reply src/shared/constants.h]: removed the todo

File: src/software/backend/simulation/simulator_ball.cpp
Comment (MathewMacDougall) [reply src/software/backend/simulation/simulator_ball.cpp]: Because it's a static variable it needs to be instantiated in a `.cpp` file. Even in the header it doesn't get default constructed without being explicitly initialized like this. (it is not hiding, it is initializing).

File: src/software/backend/simulation/physics/box2d_util.h
Comment (garethellis0) [inline src/software/backend/simulation/physics/box2d_util.h]: missing newline at EOF

File: src/software/backend/simulation/simulator_ball.cpp
Comment (garethellis0) [reply src/software/backend/simulation/simulator_ball.cpp]: Are we making this static so that we can get it's methods for passing into firmware? Could we wrap the methods in the same way that we "wrap" the robot methods so that we don't need static variables?

If we don't end up doing the above, we should at least have the constructor throw an exception if we try to create more then one instance of this class.

File: src/software/backend/simulation/simulator_ball.h
Comment (garethellis0) [inline src/software/backend/simulation/simulator_ball.h]: missing newline at EOF

File: src/software/backend/simulation/simulator_robot.cpp
Comment (garethellis0) [inline src/software/backend/simulation/simulator_robot.cpp]: Just as a note, currently we rely on the user to destruct each object they create, ie. `FirmwareRobot_t` does not take ownership of the objects you pass to it. We can look at changing this, but let's just make sure we're making an explicit choice.

File: src/software/backend/simulation/simulator_robot.cpp
Comment (garethellis0) [inline src/software/backend/simulation/simulator_robot.cpp:307]: clean, nice use of the standard library :smile:

File: src/software/backend/simulation/simulator_robot.h
Comment (garethellis0) [inline src/software/backend/simulation/simulator_robot.h:84]: I would _really_ prefer that we don't return raw pointers like this. Perhaps we could store the robots internally in a static table?

File: src/software/backend/simulation/simulator_robot.h
Comment (garethellis0) [inline src/software/backend/simulation/simulator_robot.h]: We could call this `SimulatorRobotSingleton` to indicate that there is (effectively) only ever one instance?

File: src/software/backend/simulation/simulator_robot.h
Comment (garethellis0) [inline src/software/backend/simulation/simulator_robot.h:200]: we should be able to move all the getters to be `private`?

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Super sorry, thought I had reviewed this days ago. Only larger thing here is how we manage raw pointers, see my comments in `SimulatorRobot` for that.

File: src/software/backend/simulation/physics/box2d_util.h
Comment (MathewMacDougall) [reply src/software/backend/simulation/physics/box2d_util.h]: fixed

File: src/software/backend/simulation/simulator_ball.h
Comment (MathewMacDougall) [reply src/software/backend/simulation/simulator_ball.h]: fixed

File: src/software/backend/simulation/simulator_robot.cpp
Comment (MathewMacDougall) [reply src/software/backend/simulation/simulator_robot.cpp]: Ah right. IIRC you were going to change that in a future PR, so that the object takes ownership of anything injected to it? That would make things here a lot easier. I can make an explicit note / ticket to change it or look into it again in the future. Should be ok to leak here for the very short term.

Comment (MathewMacDougall) [reply src/software/backend/simulation/simulator_robot.cpp]: made #1128

File: src/software/backend/simulation/simulator_robot.h
Comment (MathewMacDougall) [reply src/software/backend/simulation/simulator_robot.h:84]: Not sure what you mean about storing internally in a static table (don't see how that solves returning pointers).

I realized I can return a `unique_ptr` here instead since this isn't actual firmware, but gonna hold off on changes until I understand the table idea.

File: src/software/backend/simulation/simulator_robot.h
Comment (MathewMacDougall) [reply src/software/backend/simulation/simulator_robot.h]: changed for robot and ball

File: src/software/backend/simulation/simulator_robot.h
Comment (MathewMacDougall) [reply src/software/backend/simulation/simulator_robot.h:200]: Yup, changed for robot and ball

File: src/software/backend/simulation/simulator_ball.cpp
Comment (MathewMacDougall) [reply src/software/backend/simulation/simulator_ball.cpp]: > Are we making this static so that we can get it's methods for passing into firmware? Could we wrap the methods in the same way that we "wrap" the robot methods so that we don't need static variables?

It's static mostly because everything else in the singleton class is, and I'm assuming the user doesn't need to actually create an instance of the singleton. It's more clear if it's all static.

It's used in the implementation of each static function that will be bound to firmware. The robot class also uses static variables this way, so I don't understand what you mean by "wrapping" again.

> If we don't end up doing the above, we should at least have the constructor throw an exception if we try to create more then one instance of this class.

This is also a "singleton" controller class. It doesn't get instantiated / created. It's totally legal to call `setPhysicsBall` multiple times if we ever wanted to change what ball is being used / controlled.

Comment (jonathanlew) [reply src/software/backend/simulation/simulator_ball.cpp]: It doesn't look like an initialization cuz the type is being re-declared but I'm ok with it if that's how we want to do singletons.

File: src/software/backend/simulation/physics/box2d_util.cpp
Comment (jonathanlew) [inline src/software/backend/simulation/physics/box2d_util.cpp:36]: I do wonder if we should share this geom util with the rest of the code base 🤔 We're duplicating some functions, but are simulator objects different enough that it warrants it? Alternatively box2d can call geom functions

File: src/software/backend/simulation/simulator_ball.cpp
Comment (jonathanlew) [inline src/software/backend/simulation/simulator_ball.cpp:3]: Maybe I'm misunderstanding, but why are we declaring the type again?
```suggestion
SimulatorBallSingleton::physics_ball_weak_ptr =
```

**Review by jonathanlew — CHANGES_REQUESTED** [review]
Nothing much to add other than https://github.com/UBC-Thunderbots/Software/pull/1117#discussion_r368787668.

File: src/software/backend/simulation/simulator_robot.h
Comment (garethellis0) [reply src/software/backend/simulation/simulator_robot.h:84]: Sorry, I think I misread some things here, returning a `unique_ptr` should be fine.

File: src/software/backend/simulation/simulator_ball.cpp
Comment (garethellis0) [reply src/software/backend/simulation/simulator_ball.cpp]: Yeah, after looking at this again, I'm fine with this as-is.

File: src/software/backend/simulation/simulator_robot.h
Comment (MathewMacDougall) [reply src/software/backend/simulation/simulator_robot.h:84]: Changed to return a `unique_ptr`. I had to define a `Deleter` struct, so please take a look at that.

File: src/software/backend/simulation/simulator_ball.cpp
Comment (MathewMacDougall) [reply src/software/backend/simulation/simulator_ball.cpp:3]: That's just how C++ likes static variables initialized. Without the type I get the following error

> software/backend/simulation/simulator_ball.cpp:3:25: error: C++ requires a type specifier for all declarations
SimulatorBallSingleton::physics_ball_weak_ptr =

File: src/software/backend/simulation/physics/box2d_util.cpp
Comment (MathewMacDougall) [reply src/software/backend/simulation/physics/box2d_util.cpp:36]: you're right about duplication, I wrote this function in particular before a bunch of geom changes got merged.

IMO we should keep this geometry util separate from the main one as its very specific to box2d, and are mostly "wrapper" functions. We don't want general geom to depend on box2d.

I agree box2d should call our own geom functions rather than re-implementing them whenever possible. I've opened #1131 to use our own geometry once `Polygon::area` is implemented.

Comment (jonathanlew) [reply src/software/backend/simulation/physics/box2d_util.cpp:36]: #1131

**Review by jonathanlew — DISMISSED** [review]
LGTM! :+1:

File: src/software/backend/simulation/simulator_robot.h
Comment (garethellis0) [reply src/software/backend/simulation/simulator_robot.h:84]: Oh very neat!

**Review by garethellis0 — DISMISSED** [review]
LGTM :+1: :tada:

**Review by jared-paul — DISMISSED** [review]
LGTM! :+1:

**Comment by MathewMacDougall** [comment]
Added some more functions to work with updated firmware. Need quick re-review.

**Review by garethellis0 — APPROVED** [review]

**Review by jonathanlew — APPROVED** [review]
Re-approving

**Pull request merged**
